core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -249,14 +249,57 @@ class RequestSendThread(val controllerId: Int,
 
   private val socketTimeoutMs = config.controllerSocketTimeoutMs
 
+  private val controllerRequestMerger = new ControllerRequestMerger
+
+  private var firstUpdateMetadataWithPartitionsSent = false
+
+  def backoff(): Unit = pause(100, TimeUnit.MILLISECONDS)
+
   override def doWork(): Unit = {
+    val (requestBuilder, callback) = nextRequestAndCallback()
+    sendAndReceive(requestBuilder, callback)
+  }
 
-    def backoff(): Unit = pause(100, TimeUnit.MILLISECONDS)
+  private def nextRequestAndCallback(): (AbstractControlRequest.Builder[_ <: AbstractControlRequest], AbstractResponse => Unit) = {
+    if ((!config.liCombinedControlRequestEnable) || (!firstUpdateMetadataWithPartitionsSent)) {
+      val QueueItem(apiKey, requestBuilder, callback, enqueueTimeMs) = queue.take()
+      updateMetrics(apiKey, enqueueTimeMs)
+      (requestBuilder, callback)
+    } else {
+      // Only start the merging logic after the first UpdateMetadata request with partitions,
+      // since the first UpdateMetadata request with partitions may contain hundreds of thousands of partitions,
+      // and thus needs to be cached and shared by all brokers in order to prevent OOM
+
+      // there are 4 cases regarding the state of the queue and the controllerRequestMerger
+      // case 1: queue not empty, merger not empty {action: merge and send first merged request}
+      // case 2: queue not empty, merger empty {action: merge and send first merged request}
+      // case 3: queue empty, merger not empty {action: send latest merged request}
+      // case 4: queue empty, merger empty {action: block and wait until queue becomes non-empty, then transition to case 1 or 3}
+
+      // handle case 4 first
+      if (!controllerRequestMerger.hasPendingRequests()) {
+        val QueueItem(apiKey, requestBuilder, callback, enqueueTimeMs) = queue.take()
+        mergeControlRequest(enqueueTimeMs, apiKey, requestBuilder, callback)
+      }
 
-    val QueueItem(apiKey, requestBuilder, callback, enqueueTimeMs) = queue.take()
-    var queueTimeMs = time.milliseconds() - enqueueTimeMs
+      // now we are guaranteed that the controllerRequestMerger is not empty (case 1 or 3)
+      // drain the queue until the queue is empty
+      // one concurrent access case considering the producer of the queue:
+      // an item is put to the queue right after the condition check below.
+      // That behavior does not change correctness since the inserted item will be picked up in the next round
+      while (!queue.isEmpty) {
+        val QueueItem(apiKey, requestBuilder, callback, enqueueTimeMs) = queue.take()
+        mergeControlRequest(enqueueTimeMs, apiKey, requestBuilder, callback)
```

Comment:
we don't have any cap/limit on how many requests can be merged.  Is it possible that merged requests go unbound, 1) caused some mem pressure on receiving broker 2) causing any unexpected long delay since requests can be enqueued all the time.

Reply:
We talked offline and have agreed to evaluate this using stress testing in the cert cluster.